New setting to disable keeping raw request body in RAM

bigpresh opened the pull request:
Right.  Back in PR #1134, I stopped us from storing the raw request body in RAM (in the `Dancer::Request` object's `body` attribute) to avoid massive RAM usage for large requests - e.g. large file uploads, etc.

It replaced `request->body` with a convenience method that, if called, grabbed the filehandle that `HTTP::Body->body` gives you, read the content from it and returned it.

However, as per issue #1140, that caused unpleasant surprises - as `HTTP::Body` only stores the body into a filehandle in some cases, not all.

In #1153 I worked on replacing our use of HTTP::Body with a bundled version with some changes to make it more predictable - but I don't really want us to have to do that, and it still fell short of sanity for some cases, e.g. multipart file uploads.

I think this is the best solution - revert out the previous stuff, and store the request body in RAM as per the expected behaviour before #1134 (pre 1.3143) by default, to avoid surprises, but add a new config setting, `raw_request_body_in_ram`, which controls whether or not we keep the raw (unparsed) request body in `Dancer::Request->body` or not.  Users who want to avoid excessive RAM usage for large requests, and do not need access to the raw request body, should be able to set this to a false value.

bigpresh committed:
Revert "Merge branch 'bigpresh/bundle_http_body' into bigpresh/1140"

This reverts commit 7a33f70e59eab70c1bd72d75bf91aa0c84ced24e, reversing
changes made to 43bf41bc3084f5cde51ac0a9ef8b76cf969b76b0.

I'd been thinking about bundling a version of HTTP::Body that changes the
handling of request bodies, but it's not a good idea.

Instead, I'll add a config option to allow request->body to keep the raw request
body in RAM as it used to, for anyone who relies on that behaviour, and not
do so for everyone else where memory usage is important.
Conflicts:
	dist.ini
	lib/Dancer/HTTP/Body.pm
	lib/Dancer/HTTP/Body/MultiPart.pm
	lib/Dancer/HTTP/Body/OctetStream.pm
	lib/Dancer/HTTP/Body/UrlEncoded.pm
	lib/Dancer/HTTP/Body/XForms.pm
	lib/Dancer/HTTP/Body/XFormsMultipart.pm

Conflicts:
	dist.ini
	lib/Dancer/HTTP/Body.pm
	lib/Dancer/HTTP/Body/MultiPart.pm
	lib/Dancer/HTTP/Body/OctetStream.pm
	lib/Dancer/HTTP/Body/UrlEncoded.pm
	lib/Dancer/HTTP/Body/XForms.pm
	lib/Dancer/HTTP/Body/XFormsMultipart.pm

Grinnz committed:
Test newer perls in travis, and don't require test deps for dzil

As List::Lazy is only a test dep, this should fix the current failure back to 5.14.

bigpresh committed:
Remove bundled HTTP::Body from MANIFEST

bigpresh committed:
Go back to using HTTP::Body, not bundled copy, here.

bigpresh committed:
Remove compiliation tests for bundled HTTP::Body

Conflicts:
	t/00-compile.t

bigpresh committed:
Missed another use of previously-bundled H::B

bigpresh committed:
Compiling would be beneficial, I think.

bigpresh committed:
Default to the old, store raw body in RAM.

Also, only check the setting when we begin parsing, rather than calling
setting() for every single chunk of data we read from the buffer.

I'm not 100% certain, but I think that the safest bet is to to make
request->body work as expected unless the user sets the config setting to say
"don't store the request body in RAM", rather than change to it not working /
working unpredictably unless you tell it to store it for you.

bigpresh committed:
Store the raw request body in new_for_request too.

bigpresh committed:
Duh, update test count

bigpresh committed:
Tests for new raw_request_body_in_ram setting.

Ensure that it behaves as we've documented that it will.

Conflicts:
	t/02_request/07_raw_data.t

bigpresh committed:
Default request->body to an empty string.

In keeping with prior behaviour.

cromedome commented on dist.ini:
```diff
@@ -22,7 +22,10 @@ authority=cpan:SUKRIA
 
 [Prereqs / RuntimeRecommends ]
 YAML = 0
+<<<<<<< HEAD
```
Need to resolve this merge conflict.

cromedome commented on dist.ini:
```diff
@@ -31,9 +34,13 @@ YAML::XS = 0
 JSON = 2.90
 HTTP::Tiny = 0.014 ; for get/post/post_form
 HTTP::CookieJar = 0.008
+<<<<<<< HEAD
```
This one too.

cromedome commented:
I don't see anything wrong aside from the merge conflict in `dist.ini`. I'd like to poke at it a bit more, and will probably run it on one of my test servers at work for a day or two. I think I'll be ok to :+1: this though.

bigpresh committed:
Fix merge conflict.

(Not quite sure how I ended up committing this with merge markers in...)

bigpresh commented:
Finally found time for this.  If it passes its tests, I hope to merge it and get a dev release out tomorrow, give it a whirl, then get a new D1 stable release out in the coming week, as it's way overdue, and I've been sucking at taking care of D1.

bigpresh commented:
Curious, I don't get the failure about trying to add t/00_compile.t multiple times locally.

bigpresh committed:
Merge branch 'devel' into bigpresh/unbundle_http_body_make_request_body_optional_2

Conflicts:
	dist.ini

bigpresh committed:
We don't need this, it gets generated by a dzil plugin

Dist::Zilla::Plugin::Test::Compile creates this for us, we don't need to have it
in the repo too, otherwise builds fail with:

```
[DZ] attempt to add t/00-compile.t multiple times; added by: filename set by
GatherDir (Dist::Zilla::Plugin::GatherDir line 225); encoded_content added by
@Filter/GatherDir (Dist::Zilla::Plugin::GatherDir line 226); content added by
@Filter/Test::Compile (Dist::Zilla::Plugin::Test::Compile line 169); content set
by Test::Compile (Dist::Zilla::Plugin::Test::Compile line 205)
aborting; duplicate files would be produced at
/home/travis/perl5/perlbrew/perls/5.26.2/lib/site_perl/5.26.2/Dist/Zilla/App/Command/smoke.pm
line 64.
```

bigpresh committed:
Test that the setting actually changed.

bigpresh committed:
There's one more test now.

bigpresh committed:
Tests fail listening on 127.0.0.10.

This had previously been done in PR #1150 to avoid occasional test failures
where the port that Test::TCP picked to listen on ended up in use before it
tried to bind - but that didn't work for *BSD boxes, or, possibly even, any
non-Linux boxes.

It was reverted, but is still here.  Hmm.

bigpresh committed:
I added a new route to test app, update this test
